### DISCUSSION TITLE
exif のデータの持ち方

### DISCUSSION THREAD

**Discussion by takurinton** [jphacks/F_2111#33] [General]
# 取れる情報

#19 
#32 

# 考えること

データの持ち方、json の構造

**Comment by takurinton** [comment]
cc: @tetsuzawa @masibw @cn-eveer

**Reply by tetsuzawa** [reply]
スレッド生やしてディスカッションできる感じか

**Comment by tetsuzawa** [comment]
DBのスキーマのスレッド

**Reply by tetsuzawa** [reply]
これらのデータを仮として考える

```
GPSInfo （&{141.56765833333333 42.88068888888889 39}）
Make: Apple, string
Model: iPhone SE (2nd generation), string
DateTime: 2021:10:23 01:14:21, string
Flash: [16], []uint16
FocalLength: [{399 100}], []exifcommon.Rational
FNumber: [{9 5}], []exifcommon.Rational
ISOSpeedRatings: [320], []uint16
ShutterSpeedValue: [{108417 18419}], []exifcommon.SignedRational
LensModel: iPhone SE (2nd generation) back camera 3.99mm f/1.8, string
GPSImgDirectionRef: T, string
GPSImgDirection: [{228145 5478}], []exifcommon.Rational
```

**Reply by tetsuzawa** [reply]
有理数の扱いについて

DBに保存するときのメリットデメリット

- 有理数として扱う
   - ⭕ 精度落ちしない
   - ⭕ EXIFのデータを加工せずに扱える
   - ❌ 比較し辛い（毎回算術が必要）
      - 分子カラムと分母カラム作る？
      - "分子,分母"みたいに保存する？SQLで比較できなくなるけど
- FLOATとして扱う
  - 🔺 気になるほどの精度落ちはしない（緯度経度とか10^-10レベルで桁落ちしてても問題なさそう？）
  - ❌ 比較が曖昧になる（F=1.8）で検索しても引っかかるとは限らない
- DECIMALとして扱う
  - ⭕ 比較しやすい
  - 🔺 精度を考慮しなきゃいけなくてめんどい

以上を踏まえるとDECIMALがよさそう

**Reply by tetsuzawa** [reply]
EXIFテーブルを別に分けるか？について


必ず写真と1対1に紐づく情報なので分ける必要はなさそう
JOINする必要も無くなるし

**Reply by takurinton** [reply]
> EXIFテーブルを別に分けるか？について

思ったよりカラム多くないので1テーブルで良いと思います〜。
そうね、必ず写真と1対1に紐づく。

**Reply by takurinton** [reply]
有理数の扱いはここら辺の話か
https://github.com/jphacks/F_2111/issues/19#issuecomment-950267151

**Reply by tetsuzawa** [reply]
https://www.dinop.com/vc/exif04.html
> 0x9201　シャッタースピード: ShutterSpeedValue                SRATIONAL
> 　　シャッタースピードを示す。単位はAPEX。
> 　　APEX値から実際のシャッタースピードに変換するには
> 　　　　APEX値 ＝ - log2(露出時間)
> 　　という式を解く。
> 　　また、Exif Version 2.1には以下の対応表が規定されている。
> 　　　　-5 : 30sec
> 　　　　-4 : 15sec
> 　　　　-3 :  8sec
> 　　　　-2 :  4sec
> 　　　　-1 :  2sec
> 　　　　 0 :  1sec
> 　　　　 1 : 1/2sec
> 　　　　 2 : 1/4sec
> 　　　　 3 : 1/8sec
> 　　　　 4 : 1/15sec
> 　　　　 5 : 1/30sec
> 　　　　 6 : 1/60sec
> 　　　　 7 : 1/125sec
> 　　　　 8 : 1/250sec
> 　　　　 9 : 1/500sec
> 　　　　10 : 1/1000sec
> 　　　　11 : 1/2000sec


シャッタースピードの扱いが面倒そう
有理数で計算できるのはAPEX値
DBで検索するときは秒単位で扱いたいからこれもDECIMAL型にするか


少数点以下は15桁くらいあれば良さそう（ハイスピードカメラで10億コマ/秒とかのシャッタースピードだから）
整数部は6桁位あればよさそう（1日でシャッター切っても86400秒だから）
なのでDECIMAL(21,15)でいいかな

**Reply by tetsuzawa** [reply]
シャッタースピード1/30secの設定でEXIFにもピッタリした値が入ってくるかはわからないから検索機能作るときは気にする必要がありそう
範囲検索に読み替えるとか

**Reply by tetsuzawa** [reply]
GPSImgDirectionは方位、1度単位の精度で良さそう

**Reply by tetsuzawa** [reply]
GPSInfoの精度について
北緯36度では緯度or経度が1μ度（0.0000001度）ずれると1cmくらいの距離が変わるらしい
1cm単位の精度があれば流石に良さそう

**Reply by tetsuzawa** [reply]
ほんとにそんな精度いるか？

**Reply by tetsuzawa** [reply]
まあいいか

**Reply by tetsuzawa** [reply]
度数法の最大値は360なのでDECIMAL(10,7)でよさそう

**Reply by tetsuzawa** [reply]
https://www.cipa.jp/std/documents/j/DC-008-2012_J_C.pdf

EXIFの情報がかなり詳細に書いてある

**Reply by tetsuzawa** [reply]
これ真面目に対応したらカオスになるな
カメラ業界の皆様ご愁傷様です

**Reply by tetsuzawa** [reply]
@takurinton @masibw @cn-eveer 
以上を踏まえてこんなテーブル定義でいかがでしょうか

```sql
CREATE TABLE IF NOT EXISTS `photos`
(
    `id`                    CHAR(36) NOT NULL,
    `url`                   TEXT     NOT NULL,
    `title`                 TEXT     NOT NULL,
    `description`           TEXT,
    `make`                  TEXT,
    `model`                 TEXT,
    `flash`                 INT UNSIGNED,
    `focal_length`          DECIMAL(7, 2),
    `iso_speed_ratings`     INT UNSIGNED,
    `shutter_speed_value`   decimal(21, 15),
    `lens_model`            TEXT,
    `latitude`              DECIMAL(10, 7),
    `longitude`             DECIMAL(10, 7),
    `altitude`              INT UNSIGNED,
    `gps_img_direction_ref` CHAR(1),
    `gps_img_direction`     INT,
    `exif_timestamp`        TIMESTAMP,
    `created_at`            DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    `updated_at`            DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
    PRIMARY KEY (`ID`)
) ENGINE = InnoDB
  DEFAULT CHARSET = utf8mb4
  COLLATE = utf8mb4_unicode_ci;

```

**Reply by takurinton** [reply]
確認した。
めちゃくちゃ調べたりしてくれててありがたい。🙇‍♂️
テーブル設計良さそう！

**Reply by tetsuzawa** [reply]
ALTER TABLEはhttps://github.com/jphacks/F_2111/pull/18 をマージした後にやる

**Reply by masibw** [reply]
https://photohito.com/qa/
はExifを全て保存しておいて使うやつだけ抜き出してる感じがありますね...
データ量多いので気になりますが，rawExifみたいなblob or varcharのフィールドで生データ保存もありかも...?(まぁあんまり既存データからのマイグレーションは考えなくても良いので，いらない気もする

**Reply by masibw** [reply]
貼ってくださった資料によると` `iso_speed_ratings`     INT UNSIGNED,`のところは`PhotographicSensitivity`がexifでの名前な気がしますね

ただ，実際のフィールド名は`iso_speed_ratings`なのでどっちに合わせましょうか:thinking_face:

**Reply by masibw** [reply]
`gps_img_direction`    って`INT`でいいんですっけ？
0.00~359.99までなので小数考慮が必要そうな気がします

**Reply by masibw** [reply]
`exif_timestamp`  がexifのどれのことかわからなかったんですが， 撮った日時なら`DateTimeOriginal`を使えば良さそう？

**Reply by tetsuzawa** [reply]
> 貼ってくださった資料によると` `iso_speed_ratings` INT UNSIGNED,`のところは`PhotographicSensitivity`がexifでの名前な気がしますね
> 
> ただ，実際のフィールド名は`iso_speed_ratings`なのでどっちに合わせましょうか:thinking_face:

なるほど
であればexifに合わせて`PhotographicSensitivity`のほうがいい気がします。

**Reply by tetsuzawa** [reply]
> gps_img_direction ってINTでいいんですっけ？
> 0.00~359.99までなので小数考慮が必要そうな気がします

INTにした記憶がないのでDECIMALのほうが良さそうですね
自分の間違いかと！

**Reply by tetsuzawa** [reply]
> `exif_timestamp` がexifのどれのことかわからなかったんですが， 撮った日時なら`DateTimeOriginal`を使えば良さそう？

`DateTimeOriginal`のことでした！
カラム名もそっちに寄せて良さそうです

**Reply by masibw** [reply]
あざます！

**Comment by masibw** [comment]
正確に真北と磁北を計算しようとすると撮った写真の日時も関係あるのか？わからぬ
https://github.com/naturalatlas/geomagnetism

**Reply by tetsuzawa** [reply]
https://www.gsi.go.jp/buturisokuchi/menu01_index.html

> 地磁気の変化
> 地磁気は、時間によって変化します。地磁気は、擾乱のない日では、１日周期で規則的な変化を繰り返しています。時間変化は、測地観測所における地磁気連続観測毎分値グラフ、地球電磁気連続観測毎分値グラフから見ることができます。また、地磁気は、長期にわたってゆったりと変化しています。これは、永年変化と呼ばれています。

**Reply by tetsuzawa** [reply]
らしい
